helix-term/src/ui/editor.rs

```diff
@@ -231,7 +231,45 @@ impl EditorView {
             }
         }
 
-        // render selections
+        // render gutters
+
+        let linenr: Style = theme.get("ui.linenr");
+        let warning: Style = theme.get("warning");
+        let error: Style = theme.get("error");
+        let info: Style = theme.get("info");
+        let hint: Style = theme.get("hint");
+
+        for (i, line) in (view.first_line..last_line).enumerate() {
+            use helix_core::diagnostic::Severity;
+            if let Some(diagnostic) = doc.diagnostics().iter().find(|d| d.line == line) {
+                surface.set_stringn(
+                    viewport.x - OFFSET,
+                    viewport.y + i as u16,
+                    "●",
+                    1,
+                    match diagnostic.severity {
+                        Some(Severity::Error) => error,
+                        Some(Severity::Warning) | None => warning,
+                        Some(Severity::Info) => info,
+                        Some(Severity::Hint) => hint,
+                    },
+                );
+            }
+
+            // line numbers having selections are rendered differently
+            surface.set_stringn(
+                viewport.x + 1 - OFFSET,
+                viewport.y + i as u16,
+                format!("{:>5}", line + 1),
+                5,
+                linenr,
+            );
+        }
```

Comment:
Can we iterate diagnostics once only and collect all the lines into a single `HashSet` so if we have lots of diagnostics it won't be a `O(N^2)` loop?

Reply:
Yeah this needs to be solved as a part of the highlight events, the current implementation in an after-pass is a hack

Reply:
> Yeah this needs to be solved as a part of the highlight events, the current implementation in an after-pass is a hack

Are you working on the highlight events new implementation or do you want someone else to work on it, maybe open an issue for it?

Reply:
I'm trying to fix that as a part of the termwiz port, but no promises yet